### PR TITLE
NIFI-13427: Python API extended to support source processors

### PR DESCRIPTION
**Pull request by pgyori**
# Summary

Extends the NiFi Python API so that one can create Python processors that have no incoming connections but can create FlowFiles themselves.

[NIFI-13427](https://issues.apache.org/jira/browse/NIFI-13427)

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [x] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [x] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [x] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [ ] Pull Request based on current revision of the `main` branch
- [x] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [x] Build completed using `mvn clean install -P contrib-check`
  - [x] JDK 21

### Licensing

- [ ] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [ ] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [x] Documentation formatting appears as expected in rendered files

### CODE REVIEW

File: nifi-extension-bundles/nifi-py4j-bundle/nifi-python-extension-api/src/main/python/src/nifiapi/flowfilesource.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+    def getAttributes(self):
+        if self.attributes is None:
+            return None
+
+        map = JvmHolder.jvm.java.util.HashMap()
+        for key, value in self.attributes.items():
+            map.put(key, value)
+
+        return map
```
Comment (lordgamez) [inline nifi-extension-bundles/nifi-py4j-bundle/nifi-python-extension-api/src/main/python/src/nifiapi/flowfilesource.py:65]: It seems that the file endings for new files are inconsistent. Some files have file ending newlines some files don't. I think all of these new files should end with newlines.

**Review by exceptionfactory — COMMENTED** [review]
Great work on this new Python Processor interface @pgyori! The system test is very helpful as well.

If you can address the comments from @lordgamez, this looks like it should be ready to go today.

**Comment by exceptionfactory** [comment]
> Hey @pgyori 👋 Just out of curiosity - have you thought about offering a version of the `FlowFileSourceProxy` for processors that are **not** safe for concurrent execution?
> 
> Currently there is no way of setting annotations such as `@TriggerSerially `from the Python side - most regular processors are fine with 'Concurrent Tasks' > 1 but I have several source processors I would like to develop where running multiple concurrent tasks would be rather ... problematic 😇

Thanks for raising this question @henrikjohansen. This is probably worth considering in a discussion in Jira, where it would be helpful to outline the type of Processor to be developed. It is worth noting that a large number of the `List` Processors in Java have the `TriggerSerially` annotation, so it might even make sense to make that the standard behavior. However, I think this can be considered as a follow-on effort, as the implications may be a bit different given the relationship between Java Threads and Python processes.

File: nifi-extension-bundles/nifi-py4j-bundle/nifi-py4j-integration-tests/src/test/java/org.apache.nifi.py4j/PythonControllerInteractionIT.java [hunk 2]
```diff
@@ -567,6 +567,28 @@ public void testRouteToFailureWithAttributes() {
         out.assertAttributeEquals("failureReason", "Intentional failure of unit test");
     }
 
+    @Test
+    public void testCreateFlowFile() throws IOException {
+        final String processorName = "CreateFlowFile";
+        final String propertyName = "FlowFile Contents";
+        final String relationship1 = "success";
+        final String relationship2 = "multiline";
```
[the remark this answers is not in the document]
Comment (pgyori) [reply nifi-extension-bundles/nifi-py4j-bundle/nifi-py4j-integration-tests/src/test/java/org.apache.nifi.py4j/PythonControllerInteractionIT.java:575]: You are right, thanks! Renaming.

File: nifi-extension-bundles/nifi-py4j-bundle/nifi-python-extension-api/src/main/python/src/nifiapi/flowfilesource.py (hunk 1 above)
Comment (pgyori) [reply nifi-extension-bundles/nifi-py4j-bundle/nifi-python-extension-api/src/main/python/src/nifiapi/flowfilesource.py:65]: Fair point, I did not realize that. Thank you!

**Comment by pgyori** [comment]
> Hey @pgyori 👋 Just out of curiosity - have you thought about offering a version of the `FlowFileSourceProxy` for processors that are **not** safe for concurrent execution?
> 
> Currently there is no way of setting annotations such as `@TriggerSerially `from the Python side - most regular processors are fine with 'Concurrent Tasks' > 1 but I have several source processors I would like to develop where running multiple concurrent tasks would be rather ... problematic 😇

Hi @henrikjohansen ,
That is a very interesting - and challenging - problem. Thank you for asking that! I agree with @exceptionfactory , let's open a separate Jira ticket for that and discuss it further.

**Comment by pgyori** [comment]
Thank you @lordgamez , @henrikjohansen , @exceptionfactory ! 
I appreciate the feedback. I pushed a commit addressing the review comments.

**Review by lordgamez — APPROVED** [review]
LGTM

**Review by exceptionfactory — APPROVED** [review]
Thanks again @pgyori! +1 merging

**Pull request closed without merging**